File: algorithmSolution/leetCodeC++/002_Add Two Numbers.cpp
/*   DwyaneTalk@gmail.com
 *   by lutao
 *   leetCodeC++	: Add Two Numbers
 *					: https://leetcode.com/problems/add-two-numbers/
*/
#include <iostream>
#include <algorithm>
#include <vector>
#include <map>
using namespace std;


 struct ListNode {
     int val;
     ListNode *next;
     ListNode(int x) : val(x), next(NULL) {}
 };
 
 ListNode* initList(int array[], int arrSize) {
	 if(!array || !arrSize)	return NULL;
	 ListNode* ans = new ListNode(array[0]), *ptr = ans;
	 for(int i = 1; i < arrSize; ++i) {
		ptr->next = new ListNode(array[i]);
		ptr = ptr->next;
	 }
	 return ans;
 }

 void showList(ListNode* list) {
	 cout << "list: ";
	 while(list) {
		cout << list->val << " ";
		list = list->next;
	 }
	 cout << endl;
 }

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        if(!l1)	return l2;
		if(!l2)	return l1;
		int sum = l1->val + l2->val;
		int carryBit = sum >= 10;
		ListNode* ans = new ListNode(sum % 10), *ptr = ans;
		l1 = l1->next;
		l2 = l2->next;
		while(l1 && l2) {
			sum = l1->val + l2->val + carryBit;
			carryBit = sum >= 10;
			ptr->next = new ListNode(sum % 10);
			ptr = ptr->next;
			l1 = l1->next;
			l2 = l2->next;
		}
		while(l1) {
			sum = l1->val + carryBit;
			carryBit = sum >= 10;
			ptr->next = new ListNode(sum % 10);
			ptr = ptr->next;
			l1 = l1->next;
		}
		while(l2) {
			sum = l2->val + carryBit;
			carryBit = sum >= 10;
			ptr->next = new ListNode(sum % 10);
			ptr = ptr->next;
			l2 = l2->next;
		}
		if(carryBit) {
			ptr->next = new ListNode(1);
		}
		return ans;
    }
};

File: algorithmSolution/leetCodePython/002_Add Two Numbers.py
'''
DwyaneTalk@gmail.com
https://leetcode.com/problems/add-two-numbers/
'''

def initList(nums):
	size = len(nums)
	listNode = None
	if size < 1:
		return listNode
	else:
		listNode = ListNode(nums[0])
	curNode = listNode
	for x in range(1, size):
		curNode.next = ListNode(nums[x])
		curNode = curNode.next
	return listNode

def showList(listNode):
	while listNode is not None:
		print(listNode.val, end=' -> ')
		listNode = listNode.next
	print('NULL')

class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
	def addTwoNumbers(self, l1, l2):
		carry = 0
		newList = None
		curList = None
		while l1 is not None and l2 is not None:
			val = l1.val + l2.val + carry
			if newList is None:
				newList = ListNode(val % 10)
				curList = newList
			else:
				curList.next = ListNode(val % 10)
				curList = curList.next
			carry = int(val >= 10)
			l1 = l1.next
			l2 = l2.next
		while l1 is not None:
			val = l1.val + carry
			curList.next = ListNode(val % 10)
			curList = curList.next
			carry = int(val >= 10)
			l1 = l1.next
		while l2 is not None:
			val = l2.val + carry
			curList.next = ListNode(val % 10)
			curList = curList.next
			carry = int(val >= 10)
			l2 = l2.next
		if carry:
			curList.next = ListNode(1)
		return newList

if __name__ == '__main__':
	solution = Solution()
	listNode1 = initList([2, 4, 3])
	listNode2 = initList([5, 6, 4])
	showList(listNode1)
	showList(solution.addTwoNumbers(listNode1, listNode2))